### DISCUSSION TITLE
tutorials/shaders/using_viewport_as_texture

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#246] [User-contributed notes]
# tutorials/shaders/using_viewport_as_texture

Introduction: This tutorial will introduce you to using the SubViewport as a texture that can be applied to 3D objects. In order to do so, it will walk you through the process of making a procedura...

https://docs.godotengine.org/en/stable/tutorials/shaders/using_viewport_as_texture.html

**Comment by rybern** [comment]
This is a super helpful tutorial, but I think I can see two things to improve. Keep in mind that I'm a total beginner so I might be wrong.
1. Instead of computing the 3d noise texture in the shader, you can precompute it and pass it in as a uniform.
2. This can be done with a spacial shader in the same way, expect then you don't need to set up the SubViewport at all. Maybe I'm missing the benefit?

**Reply by tetrapod00** [reply]
1. Yep! You can precompute the noise and pass it in.
2. You're right, a viewport is not necessary to replicate this particular effect. It can be done in a single spatial shader instead. The tutorial is about using a SubViewport as a texture and happens to use this example.

At some point this page could be rewritten with an example that actually *requires* the use of a ViewportTexture. If you have any ideas, feel free to open an issue in [the docs repo](https://github.com/godotengine/godot-docs/issues).

**Comment by Lathryx** [comment]
How would this tutorial work now with the new `ORMMaterial3D`s?

**Reply by Calinou** [reply]
It works the same for the albedo part. The roughnesss part will change though, as your ViewportTexture must write its roughness to the green channel of the texture only, leaving the red/blue channel unaffected. In practice, you'll probably want the red channel to be 1.0 for every pixel (if you have no ambient occlusion map), and the blue channel to be 0.0 for every pixel (the planet is dielectric, i.e. non-metallic).

**Comment by Ducksteron** [comment]
If you just want a noise texture on a sphere (like on a skybox) you can also just use sampler3D for preprocessed 3D noise.

**Comment by C-Corpze** [comment]
I've been wondering if this tutorial can somehow be extrapolated to use something similar for skinned meshes?
I've been trying to look for a way to basically paint onto an animated character while they're moving.

I see very little to no tutorials on how to render textures onto meshes of any shape.

**Reply by Calinou** [reply]
[Decals](https://docs.godotengine.org/en/stable/tutorials/3d/using_decals.html) can achieve this if the decal is parented to a bone, but it won't actually modify texture data.

**Reply by C-Corpze** [reply]
I specifically need render targets or viewport textures because I want to sample the texture and apply blending and displacement to the mesh itself.
